### PR TITLE
Sherif akoush issue/2621/refactor env var retrieval

### PR DESCRIPTION
**Pull request by sakoush**
Consolidate the logic of getting values from enviornment variables as they were duplicated in two places. 

**What this PR does / why we need it**:

Tidy up code.

**Which issue(s) this PR fixes**:

Fixes https://github.com/SeldonIO/seldon-core/issues/2621

**Special notes for your reviewer**:

I have introduced `env_utils.py` to contain the refactored logic. Otherwise there will be circular dependencies.

**Does this PR introduce a user-facing change?**:

```release-note
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **majolo**
You can assign the PR to them by writing `/assign @majolo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: python/seldon_core/env_utils.py
Comment (RafalSkolasinski) [inline python/seldon_core/env_utils.py:1]: Could we just have these in the utils.py? Is there a particular reason to have a separate utils file/

File: python/seldon_core/metrics.py
Comment (RafalSkolasinski) [inline python/seldon_core/metrics.py:16]: we could just (if we put this into `utils)`
```
from seldon_core import utils
```

and then use `utils.get_image_name` wherever needed

Though both works and and usually over-import so does not require a change now.

File: python/seldon_core/utils.py
Comment (RafalSkolasinski) [inline python/seldon_core/utils.py:43]: Why the extra assertion? Just falling back to non-implemented seemed to work fine.

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]
Nice! Just a few questions

File: python/seldon_core/env_utils.py
Comment (sakoush) [reply python/seldon_core/env_utils.py:1]: There seems to be some circular dependencies if we do that
```
seldon_core/metrics.py:16: in <module>
    from seldon_core.utils import get_image_name, get_model_name, get_deployment_name, get_predictor_name, \
seldon_core/utils.py:16: in <module>
    from seldon_core.user_model import (
seldon_core/user_model.py:10: in <module>
    from seldon_core.metrics import SeldonMetrics, validate_metrics
```
 I have not investigated a lot into that though.  I think that having a separate module to have all env variables handling is ok. isnt it?

File: python/seldon_core/utils.py
Comment (sakoush) [reply python/seldon_core/utils.py:43]: I felt that if `ENV_MODEL_NAME` env variable is set, it should reference a valid docker image. So `ENV_MODEL_IMAGE` needs to be set as well? Perhaps to run locally this is not the case though but I am not sure.

I guess the question is really whether it is odd to return `{"proper_model_name": "non-implemented"}`? Happy to take the extra assert if it is fine for users to set one env variable and not the other.

File: python/seldon_core/env_utils.py
Comment (RafalSkolasinski) [reply python/seldon_core/env_utils.py:1]: Yeah, it's already an improvement to have them in one place so if there is no obvious solution to circular dependency let's leave it as it is :+1:

File: python/seldon_core/utils.py
Comment (RafalSkolasinski) [reply python/seldon_core/utils.py:43]: Hmm.. I guess it should be okay. In k8s both should be set by operator so won't matter much. As long as tests will pas let's leave it as it is.

File: python/seldon_core/env_utils.py
Comment (adriangonz) [inline python/seldon_core/env_utils.py:16]: typo: `get_predictor_version` instead of `get_predictior_version`

File: python/seldon_core/utils.py
Comment (adriangonz) [inline python/seldon_core/utils.py:39]: Conscious that it falls slightly outside the scope of this PR, but it would make sense to return `None` rather than `NONIMPLEMENTED_MSG`?

File: python/seldon_core/utils.py
Comment (adriangonz) [inline python/seldon_core/utils.py:355]: Nice catch :+1:

File: python/tests/test_model_microservice.py
Comment (adriangonz) [inline python/tests/test_model_microservice.py:244]: Would it make more sense to mock the environment variables rather than the methods?

Comment (adriangonz) [reply python/tests/test_model_microservice.py:244]: Having said that, I just realised that you are testing the environment as well down below so maybe you're right in that it makes more sense to mock only the methods here.

**Review by adriangonz — COMMENTED** [review]
Nice one @sakoush ! Just had a look and, other than a small typo, everything looks good :+1:

File: python/seldon_core/utils.py
Comment (RafalSkolasinski) [reply python/seldon_core/utils.py:39]: I think some of these are sometimes used for a key in dictionary. `None` cannot be a key as far as I remember.

**Comment by seldondev** [comment]
@sakoush: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e6c3ceb0521e722cb6d23533df2a330afdb3be1e | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3420-integration-wt4p8) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: python/seldon_core/env_utils.py
Comment (sakoush) [reply python/seldon_core/env_utils.py:16]: nice catch

File: python/seldon_core/utils.py
Comment (sakoush) [reply python/seldon_core/utils.py:39]: it can but if we read it back it will be "None" and not `None` so we need to convert anyway

**Comment by seldondev** [comment]
@sakoush: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e6c3ceb0521e722cb6d23533df2a330afdb3be1e | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3420-integration-wt4p8) | `/test integration`
notebooks | e6c3ceb0521e722cb6d23533df2a330afdb3be1e | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-3420-notebooks-szjmv) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@sakoush: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e6c3ceb0521e722cb6d23533df2a330afdb3be1e | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-3420-notebooks-szjmv) | `/test notebooks`
integration | c1b10013a6720c231da0183c3faa84081c41e888 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3420-integration-qnp6f) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by RafalSkolasinski** [comment]
Integration tests
```
================== 4 failed, 31 passed in 3246.50s (0:54:06) ===================
```

```
_______________________ TestPrepack.test_openapi_sklearn _______________________
[gw1] linux -- Python 3.7.10 /opt/conda/bin/python

self = <test_prepackaged_servers.TestPrepack object at 0x7f636d7f1110>
namespace = 'test-openapi-sklearn'

    def test_openapi_sklearn(self, namespace):
        spec = "../../servers/sklearnserver/samples/iris.yaml"
        retry_run(f"kubectl apply -f {spec} -n {namespace}")
        wait_for_status("sklearn", namespace)
>       wait_for_rollout("sklearn", namespace)

test_prepackaged_servers.py:250: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
seldon_e2e_utils.py:161: in wait_for_rollout
    wait_for_deployment(deployment_name, namespace, attempts, sleep)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

deployment_name = 'sklearn-default-0-classifier'
namespace = 'test-openapi-sklearn', attempts = 50, sleep = 5

    def wait_for_deployment(deployment_name, namespace, attempts=50, sleep=5):
        logging.info(f"Waiting for deployment {deployment_name}")
        for _ in range(attempts):
            ret = run(
                f"kubectl rollout status -n {namespace} deploy/{deployment_name}",
                shell=True,
            )
            if ret.returncode == 0:
                logging.info(f"Successfully waited for deployment {deployment_name}")
                break
            logging.warning(f"Unsuccessful wait command but retrying for {deployment_name}")
            time.sleep(sleep)
        assert (
            ret.returncode == 0
>       ), f"Wait for rollout of {deployment_name} failed: non-zero return code"
E       AssertionError: Wait for rollout of sklearn-default-0-classifier failed: non-zero return code

seldon_e2e_utils.py:140: AssertionError
```


```
____________________ TestPrepack.test_text_alibi_explainer _____________________
[gw1] linux -- Python 3.7.10 /opt/conda/bin/python

self = <test_prepackaged_servers.TestPrepack object at 0x7f636d674610>
namespace = 'test-text-alibi-explainer'

    def test_text_alibi_explainer(self, namespace):
        spec = "../resources/movies-text-explainer.yaml"
        retry_run(f"kubectl apply -f {spec} -n {namespace}")
        wait_for_status("movie", namespace)
>       wait_for_rollout("movie", namespace, expected_deployments=2)

test_prepackaged_servers.py:207: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
seldon_e2e_utils.py:161: in wait_for_rollout
    wait_for_deployment(deployment_name, namespace, attempts, sleep)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

deployment_name = 'movie-movies-predictor-0-classifier'
namespace = 'test-text-alibi-explainer', attempts = 50, sleep = 5

    def wait_for_deployment(deployment_name, namespace, attempts=50, sleep=5):
        logging.info(f"Waiting for deployment {deployment_name}")
        for _ in range(attempts):
            ret = run(
                f"kubectl rollout status -n {namespace} deploy/{deployment_name}",
                shell=True,
            )
            if ret.returncode == 0:
                logging.info(f"Successfully waited for deployment {deployment_name}")
                break
            logging.warning(f"Unsuccessful wait command but retrying for {deployment_name}")
            time.sleep(sleep)
        assert (
            ret.returncode == 0
>       ), f"Wait for rollout of {deployment_name} failed: non-zero return code"
E       AssertionError: Wait for rollout of movie-movies-predictor-0-classifier failed: non-zero return code

seldon_e2e_utils.py:140: AssertionError
```

```
___________________________ TestPrepack.test_sklearn ___________________________
[gw1] linux -- Python 3.7.10 /opt/conda/bin/python

self = <test_prepackaged_servers.TestPrepack object at 0x7f636d6d3f90>
namespace = 'test-sklearn'

    def test_sklearn(self, namespace):
        spec = "../../servers/sklearnserver/samples/iris.yaml"
        retry_run(f"kubectl apply -f {spec} -n {namespace}")
        wait_for_status("sklearn", namespace)
>       wait_for_rollout("sklearn", namespace)

test_prepackaged_servers.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
seldon_e2e_utils.py:161: in wait_for_rollout
    wait_for_deployment(deployment_name, namespace, attempts, sleep)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

deployment_name = 'sklearn-default-0-classifier', namespace = 'test-sklearn'
attempts = 50, sleep = 5

    def wait_for_deployment(deployment_name, namespace, attempts=50, sleep=5):
        logging.info(f"Waiting for deployment {deployment_name}")
        for _ in range(attempts):
            ret = run(
                f"kubectl rollout status -n {namespace} deploy/{deployment_name}",
                shell=True,
            )
            if ret.returncode == 0:
                logging.info(f"Successfully waited for deployment {deployment_name}")
                break
            logging.warning(f"Unsuccessful wait command but retrying for {deployment_name}")
            time.sleep(sleep)
        assert (
            ret.returncode == 0
>       ), f"Wait for rollout of {deployment_name} failed: non-zero return code"
E       AssertionError: Wait for rollout of sklearn-default-0-classifier failed: non-zero return code

seldon_e2e_utils.py:140: AssertionError
```

```
______________________ TestBatchWorker.test_batch_worker _______________________
[gw0] linux -- Python 3.7.10 /opt/conda/bin/python

self = <test_batch_processor.TestBatchWorker object at 0x7f7145b52650>
namespace = 'test-batch-worker'

    def test_batch_worker(self, namespace):
        spec = "../../servers/sklearnserver/samples/iris.yaml"
        retry_run(f"kubectl apply -f {spec} -n {namespace}")
        wait_for_status("sklearn", namespace)
>       wait_for_rollout("sklearn", namespace)

test_batch_processor.py:29: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
seldon_e2e_utils.py:161: in wait_for_rollout
    wait_for_deployment(deployment_name, namespace, attempts, sleep)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

deployment_name = 'sklearn-default-0-classifier'
namespace = 'test-batch-worker', attempts = 50, sleep = 5

    def wait_for_deployment(deployment_name, namespace, attempts=50, sleep=5):
        logging.info(f"Waiting for deployment {deployment_name}")
        for _ in range(attempts):
            ret = run(
                f"kubectl rollout status -n {namespace} deploy/{deployment_name}",
                shell=True,
            )
            if ret.returncode == 0:
                logging.info(f"Successfully waited for deployment {deployment_name}")
                break
            logging.warning(f"Unsuccessful wait command but retrying for {deployment_name}")
            time.sleep(sleep)
        assert (
            ret.returncode == 0
>       ), f"Wait for rollout of {deployment_name} failed: non-zero return code"
E       AssertionError: Wait for rollout of sklearn-default-0-classifier failed: non-zero return code

seldon_e2e_utils.py:140: AssertionError
```

**Comment by seldondev** [comment]
@sakoush: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c1b10013a6720c231da0183c3faa84081c41e888 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3420-integration-qnp6f) | `/test integration`
notebooks | c1b10013a6720c231da0183c3faa84081c41e888 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-3420-notebooks-vwxc2) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@sakoush: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c1b10013a6720c231da0183c3faa84081c41e888 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3420-integration-qnp6f) | `/test integration`
notebooks | 476a7dd9c723dd521ddd869d2760e490f4fa3830 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-3420-notebooks-525wm) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by RafalSkolasinski** [comment]
The failure is on `TestNotebooks.test_custom_metrics` so most likely flakiness :+1:

**Comment by adriangonz** [comment]
Thanks for the heads up @RafalSkolasinski! In that case, is it alright to merge this one?

**Comment by RafalSkolasinski** [comment]
Should be alright but the new tests should finish soon so we can wait a few minutes and see

**Comment by seldondev** [comment]
@sakoush: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 476a7dd9c723dd521ddd869d2760e490f4fa3830 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/io-seldon-core-pr-3420-notebooks-qmwqd) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
